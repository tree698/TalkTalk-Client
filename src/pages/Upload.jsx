import React, { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import { BsPlusLg } from 'react-icons/bs';
import { BsChevronRight } from 'react-icons/bs';
import { MdDownloading } from 'react-icons/md';
import { useApiContext } from '../context/ApiContext';
import { useNavigate } from 'react-router-dom';
import { FileUpload } from '../components/FileUpload';
import Footer from '../components/Footer';
import Navbar from '../components/Navbar';
import Banner from '../components/ui/Banner';

export default function Upload() {
  const [info, setInfo] = useState({ title: '', description: '', brush: '' });
  const [fileName, setFileName] = useState('');
  const [originalName, setOriginalName] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [success, setSuccess] = useState('');
  const [error, setError] = useState('');

  const { workService } = useApiContext();
  const navigate = useNavigate();

  useEffect(() => {
    success && toast.success(success);
  }, [success]);

  useEffect(() => {
    error && toast.error(error);
  }, [error]);

  const handlePhoto = (photo) => {
    setFileName((prev) => photo.fileName);
    setOriginalName((prev) => photo.originalName);
  };

  const handleCancel = () => {
    setFileName('');
    setOriginalName();
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    setIsLoading(true);
    workService
      .createWork(
        info.title,
        info.description,
        info.brush,
        originalName,
        fileName
      )
      .then(() => {
        setSuccess('Successfully Uploaded!');
        setTimeout(() => {
          setSuccess(undefined);
        }, 3000);
        setTimeout(() => {
          navigate('/home');
        }, 3000);
      })
      .catch((error) => setError((prev) => error.toString()))
      .finally(() => setIsLoading(false));
  };

  const handleChange = (event) => {
    const { name, value } = event.target;
    setInfo((prev) => ({ ...info, [name]: value }));
  };

  const INPUT_STYLE =
    'w-full px-2 py-3 text-xl outline-none rounded-md border border-superLightGray mb-3 placeholder:text-superLightGray placeholder:italic placeholder:text-lg';
  const FILE_UPLOAD_STYLE =
    'h-[378px] flex items-center justify-center rounded-lg text-xl border-2 border-superLightGray shadow-inner shadow-xl';
  const LABEL_STYLE = 'text-xl mb-2';

  return (
    <section className="w-full h-full flex flex-col">
      <Navbar />
      <section className="flex-1 max-w-5xl w-full mx-auto">
        <div className="flex justify-center">
          <h2 className="text-3xl font-semibold m-12">Upload your drawing</h2>
        </div>
        <div className="w-full grid grid-cols-1 md:grid-cols-2 gap-4 mb-10">
          <div className={`${FILE_UPLOAD_STYLE}`}>
            <FileUpload sendPhoto={handlePhoto} />
          </div>
          <div className={`${FILE_UPLOAD_STYLE}`}>
            {fileName ? (
              <img
                src={`${process.env.REACT_APP_BASE_URL}/uploaded_images/${fileName}`}
                alt={info.title}
                className="w-full h-full"
              />
            ) : (
              <BsPlusLg />
            )}
          </div>
        </div>
        <div>
          {fileName && (
            <div>
              {isLoading && (
                <div className="flex items-center justify-center gap-4 mt-12 text-2xl">
                  <MdDownloading className="text-3xl" />
                  <Banner text="Loading..." />
                </div>
              )}
              <button
                onClick={handleCancel}
                className="flex items-center justify-center mx-auto w-[400px] text-xl text-center py-3 m-10 border border-superLightGray rounded-2xl shadow-xl hover:scale-105 hover:shadow-xl transition-all delay-150 duration-300 ease-in-out"
              >
                <BsChevronRight className=" mr-5" />
                <p>Want you to cancel it?</p>
              </button>
            </div>
          )}
        </div>
        <form onSubmit={handleSubmit}>
          <label htmlFor="title" className={`${LABEL_STYLE}`}>
            Title
          </label>{' '}
          <br />
          <input
            type="text"
            name="title"
            value={info.title}
            onChange={handleChange}
            placeholder="Title"
            id="title"
            required
            className={`${INPUT_STYLE}`}
          />
          <br />
          <label htmlFor="brush" className={`${LABEL_STYLE}`}>
            Brush
          </label>{' '}
          <br />
          <input
            type="text"
            name="brush"
            value={info.brush}
            onChange={handleChange}
            placeholder="(Optional) Brush"
            id="brush"
            className={`${INPUT_STYLE}`}
          />
          <br />
          <label htmlFor="description" className={`${LABEL_STYLE}`}>
            Say something
          </label>
          <br />
          <textarea
            name="description"
            value={info.description}
            onChange={handleChange}
            cols="30"
            rows="3"
            placeholder="(Optional) Say something"
            id="description"
            className={`${INPUT_STYLE}`}
          ></textarea>
          <br />
          <div className="flex justify-center">
            <button className="w-[400px] py-3 px-2 my-6 text-xl font-semibold text-center text-white bg-accent rounded-2xl hover:scale-105 hover:shadow-xl transition-all delay-150 duration-300 ease-in-out">
              Submit
            </button>
          </div>
        </form>
      </section>
      <Footer />
    </section>
  );
}
