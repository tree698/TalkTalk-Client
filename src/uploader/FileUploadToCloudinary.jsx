import React, { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import Dropzone from 'react-dropzone';
import { MdDownloading } from 'react-icons/md';
import { BsCheckLg } from 'react-icons/bs';
import Banner from '../components/ui/Banner';
import { uploadImage } from './uploader';

export const FileUploadToCloudinary = ({ sendImageData }) => {
  const [success, setSuccess] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState();

  useEffect(() => {
    error && toast.success(error);
  }, [error]);

  const dropHandler = async (file) => {
    setIsLoading(true);
    uploadImage(file[0])
      .then((data) => {
        sendImageData(data);
        setIsLoading(false);
        setSuccess(true);
        setTimeout(() => {
          setSuccess(false);
        }, 3000);
      })
      .catch((error) => setError((prev) => error.toString()));
  };

  return (
    <>
      <Dropzone onDrop={dropHandler}>
        {({ getRootProps, getInputProps }) => (
          <section>
            <div {...getRootProps()}>
              <input {...getInputProps()} />
              <button>
                {isLoading && (
                  <div className="flex items-center justify-center gap-2 text-xl">
                    <MdDownloading className="text-xl" />
                    <Banner text="Loading..." />
                  </div>
                )}
                {success && (
                  <div className="flex items-center justify-center gap-2 text-xl">
                    <BsCheckLg className="text-xl" />
                    <Banner text="Your image is successfully uploaded" />
                  </div>
                )}
                {!success && !isLoading && (
                  <p>
                    Drag & Drop Here <br /> or Click Here
                  </p>
                )}
              </button>
            </div>
          </section>
        )}
      </Dropzone>
    </>
  );
};
