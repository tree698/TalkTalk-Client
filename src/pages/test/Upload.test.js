import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withApiContext, withRouter } from '../../test/utils';
import Upload from '../Upload';
import { FileUploadToCloudinary } from '../../uploader/FileUploadToCloudinary';

jest.mock('../../components/Navbar');
jest.mock('../../components/Footer');
jest.mock('../../uploader/FileUploadToCloudinary');

describe('Upload', () => {
  const fakeContext = {
    workService: {
      createWork: jest.fn(),
    },
  };

  afterEach(() => fakeContext.workService.createWork.mockReset());

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<Upload />} />),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('display input value correctly', async () => {
    render(
      withApiContext(
        withRouter(<Route path="/" element={<Upload />} />),
        fakeContext
      )
    );

    const inputElement = screen.getByRole('textbox', { name: /title/i });
    userEvent.type(inputElement, 'test');

    await waitFor(() => expect(inputElement).toHaveValue('test'));
  });
});
