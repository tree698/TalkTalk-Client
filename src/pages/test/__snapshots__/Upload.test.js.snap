// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload renders correctly 1`] = `
<section
  className="w-full h-full flex flex-col"
>
  <div
    className="w-full fixed top-0 left-0 bg-white"
  />
  <section
    className="w-full md:max-w-2xl lg:max-w-3xl mx-auto flex-1 px-3 md:px-0 pt-[66px]"
  >
    <div
      className="flex justify-center"
    >
      <h2
        className="text-lg md:text-base lg:text-xl m-6 md:m-8 lg:m-10 font-semibold"
      >
        Upload your drawing
      </h2>
    </div>
    <div
      className="w-full grid grid-cols-1 md:grid-cols-2 gap-1 md:gap-2 lg:gap-3 mb-6 md:mb-8 lg:mb-10"
    >
      <div
        className="h-[291px] md:h-[249px] text-xs md:text-sm lg:text-base flex items-center justify-center rounded-lg border border-superLightGray shadow-inner shadow-xl"
      />
      <div
        className="h-[291px] md:h-[249px] text-xs md:text-sm lg:text-base flex items-center justify-center rounded-lg border border-superLightGray shadow-inner shadow-xl"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2Z"
            fillRule="evenodd"
          />
        </svg>
      </div>
    </div>
    <div />
    <form
      onSubmit={[Function]}
    >
      <label
        className="text-xs md:text-sm lg:text-base mb-1 md:mb-2"
        htmlFor="title"
      >
        Title
      </label>
       
      <br />
      <input
        className="w-full px-1 md:px-2 py-1 md:py-2 text-xs md:text-sm lg:text-base outline-none rounded-md border border-superLightGray mb-1 md:mb-2 placeholder:text-superLightGray placeholder:italic placeholder:text-xs md:placeholder:text-sm lg:placeholder:text-base"
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
      <br />
      <label
        className="text-xs md:text-sm lg:text-base mb-1 md:mb-2"
        htmlFor="brush"
      >
        Brush
      </label>
       
      <br />
      <input
        className="w-full px-1 md:px-2 py-1 md:py-2 text-xs md:text-sm lg:text-base outline-none rounded-md border border-superLightGray mb-1 md:mb-2 placeholder:text-superLightGray placeholder:italic placeholder:text-xs md:placeholder:text-sm lg:placeholder:text-base"
        id="brush"
        name="brush"
        onChange={[Function]}
        placeholder="(Optional) Brush"
        type="text"
        value=""
      />
      <br />
      <label
        className="text-xs md:text-sm lg:text-base mb-1 md:mb-2"
        htmlFor="description"
      >
        Say something
      </label>
      <br />
      <textarea
        className="w-full px-1 md:px-2 py-1 md:py-2 text-xs md:text-sm lg:text-base outline-none rounded-md border border-superLightGray mb-1 md:mb-2 placeholder:text-superLightGray placeholder:italic placeholder:text-xs md:placeholder:text-sm lg:placeholder:text-base"
        cols="30"
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="(Optional) Say something"
        rows="3"
        value=""
      />
      <br />
      <div
        className="flex justify-center"
      >
        <button
          className="w-52 md:w-56 lg:w-60 text-xs md:text-sm lg:text-base py-1 md:py-2 px-1 md:px-2 my-2 md:my-4 lg:my-6 font-semibold text-center text-white bg-brand rounded-2xl hover:scale-105 hover:shadow-xl transition-all delay-150 duration-300 ease-in-out"
        >
          Submit
        </button>
      </div>
    </form>
  </section>
</section>
`;
